### DISCUSSION TITLE
How to remove from cstom Index

### DISCUSSION THREAD

**Discussion by vitalybrandes** [OrchardCMS/OrchardCore#9828] [Q&A]
I have custom IndexProvider that add content Item from type X to index each time new content item generated.
How can i remove it from the **Index** while that Content Item X was delete from Orchard.

**Comment by Skrypt** [comment]
It is supposed to remove them.

**Reply by vitalybrandes** [reply]
@Skrypt  Well, in [ContentItemIndex ](https://github.com/OrchardCMS/OrchardCore/blob/1ec2e9898ff37e49d4224732c8a3e0acb5ed2fa6/src/OrchardCore/OrchardCore.ContentManagement/Records/ContentItemIndex.cs)its only change the "Published" to false and not actually delete it,
i would like to clean my index table and cannot find how to.
_P.S i remember that i saw it somewhere...._

**Reply by Skrypt** [reply]
Yes, in Orchard everything is only soft deleted. It reflects what is done in the ContentItemIndex. So, for deleting completely a ContentItem we would need a "Recycle Bin" feature in OC. Which would have the functionality to remove ContentItems in the Document, ContentItemIndex and custom Indexes tables.

**Reply by vitalybrandes** [reply]
@Skrypt , yeah, soft delete is ok form me,
but why i cannot remove line from the Index table? (My index table)

**Reply by Skrypt** [reply]
So basically a view that lists all items that are Published = false and Latest = false.

**Reply by Skrypt** [reply]
I need to try but I think you have a point there.

**Reply by vitalybrandes** [reply]
So i cannot remove it from my own custom index Providor? never?
and this Line in database will stay forever?

**Reply by Skrypt** [reply]
Well, the IndexingProvider describes your index content and it's triggered on every changes made on your ContentItem. So, I need to check what it does when you delete a Content Item. But I'm pretty sure there is a way to do this.

**Reply by vitalybrandes** [reply]
> 
> 
> Well, the IndexingProvider describes you index content and it,s triggered on every changes made on your ContentItem. So, I need to check what it does when you delete a Content Item. But I'm pretty sure there is a way to do this.

I almost sure that i saw it somewhere :-)

**Comment by deanmarcussen** [comment]
when you are indexing them, you should only return an index value, if the item is Published, if you only want to track published items.

So the choice to make is in your indexing provider

**Reply by vitalybrandes** [reply]
@deanmarcussen 
I am indexing in "CustomerIndex" table customers content items.
Once customer removed, i would like to delete that line from the CustomerIndex db table.

**Reply by Skrypt** [reply]
I have something similar in my website let me try it.

**Reply by Skrypt** [reply]
I think you need to return null from the IndexingProvider when the ContentItem passed is Published = false and Latest = false but I will try it.

**Reply by vitalybrandes** [reply]
@Skrypt  this will not create the Index, i am doing it.
But how to remove the exisning intex in the table?


```
public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<CompanyByPrinterIndex>()
             .Map(async contentItem =>
             {
                 var company = contentItem.As<Company>();

                 if (company == null)
                 {
                     return null;
                 }

                 var results = new List<CompanyByPrinterIndex>();

                 //for each selected printer create index
                 foreach (string _Id in company.Printer.ContentItemIds)
                 {
                     _contentManager ??= _serviceProvider.GetRequiredService<IContentManager>();

                     var printer = await _contentManager.GetAsync(_Id);

                     results.Add(new CompanyByPrinterIndex
                     {
                         Company = company.CompanyName.Text,
                         CompanyId = company.CompanyId.Text,
                         CompanyContentItemId = company.ContentItem.ContentItemId,
                     }); ;
                 }
                 return results;
             });
        }
```

**Reply by Skrypt** [reply]
```
public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<CompanyByPrinterIndex>()
             .Map(async contentItem =>
             {
                 var company = contentItem.As<Company>();

                 if (company == null)
                 {
                     return null;
                 }

                 // Removes the index lines in the db if there are any for this ContentItem when Published = false and Latest = false
                 if(!contentItem.Published && !contentItem.Latest)
                 {
                     return null;
                 }

                 var results = new List<CompanyByPrinterIndex>();

                 //for each selected printer create index
                 foreach (string _Id in company.Printer.ContentItemIds)
                 {
                     _contentManager ??= _serviceProvider.GetRequiredService<IContentManager>();

                     var printer = await _contentManager.GetAsync(_Id);

                     results.Add(new CompanyByPrinterIndex
                     {
                         Company = company.CompanyName.Text,
                         CompanyId = company.CompanyId.Text,
                         CompanyContentItemId = company.ContentItem.ContentItemId,
                     }); ;
                 }
                 return results;
             });
        }
```

While this works it won't remove it from the LuceneIndex if you are indexing drafts. I need to fix that one.

**Reply by vitalybrandes** [reply]
@Skrypt  Perfect!
Copy your answer to new , i will mark this as solution :-)

**Reply by Skrypt** [reply]
Well let me debug this a little more I think I found an issue with this though.

**Reply by Skrypt** [reply]
Ok I tried publishing the ContentItem, then saving a draft on top of the published content item. Then deleted the content item and it removed all the lines associated with this content item in the Indexing table. So, it works. Though I've had one issue where it kept the item in the Lucene index as a draft. I need to repro that one but it basically works.

**Reply by vitalybrandes** [reply]
For me is 100% ok! 
Thanks for help!
